### ISSUE TITLE
performance_matrix_multiplication, the actual peak flops is much higher than the theoretical peak flops

### ISSUE THREAD

**Issue by Bbean** [halide/Halide#8480]
When testing performance_matrix_multiplication, the actual peak flops is much higher than the theoretical peak flops. I'd like to know why. 
I work on Intel Core i5-8500B. The theoretical peak value should be 
`6 cores × 3.0 GHz × 8 FLOPs/cycle * 2 (mul + add) = 288 GFlops. `
However, the actual peak value of the calculation in the output I got is 
`Halide: 3.631948ms, 537.558090 GFLOP/s`, which is even higher than the theoretical peak flops after turbo (4.1 GHz).
 I'd like to know where the possible problem lies. Thanks for any possible answers.

**Comment by mcourteaux** [comment]
On the [Coffee Lake block diagram](https://en.wikichip.org/wiki/intel/microarchitectures/coffee_lake), I see 2 FP FMA units:

![CoffeLake](https://en.wikichip.org/w/images/thumb/7/7e/skylake_block_diagram.svg/900px-skylake_block_diagram.svg.png)

Rerunning the math, but with 2 FP FMA's doubles the number of flops, yielding 576 GFlops, which is above what Halide reported. Halide achieves 93.3% of theoretical peak, with these calculations, at 3GHz.

**Comment by Bbean** [comment]
That's a reasonable explanation, Thank you for your reply. 93.3% is a very impressive hit.

**Comment by mcourteaux** [comment]
Ah dang, that's actually super sensible. Thanks for pointing that out. Does 93% of peak sound reasonable to you?

**Comment by abadams** [comment]
Yes. The correct inner loop for a matrix multiply is a standard known thing. You just have to tile such that things are limited by the number of fmas rather than the number of loads. Coffee Lake can do two aligned loads per cycle and two fmas per cycle, so there are lots of options that work. It was harder when you could only do one load per cycle. Given that, how close to peak you reach is largely determined by how you use the cache hierarchy. That CPU has 9 MB of L3. The two matrices involved add up to ~8MB, so they stay in L3. With any kind of sensible blocking over the i,j,k axes to get things into L2 and L1 at appropriate times, this makes it possible to keep the fma units fed. Blocking over the k axis is a bit annoying because it requires factoring the reduction.

If you make the matrices twice as large so they fall out of L3 performance is probably going to be worse. Matrix multiplication out of dram is a harder problem.

**Comment by Bbean** [comment]
I am studying this mat_mul schedule, and i really confused by the schedule order:

**Comment by Bbean** [comment]
@abadams and I found that it was really hard to understand this new schedule.
I used `compile_to_c` and may know the actual calculation process, two intermediate buffers, blocking over the k axis. But the connection between this calculation process and the schedule got me stuck. I didn't figure out what the sequence of thinking between them was.
```
out.tile...
matrix_mul.update...
intm.compute_at...
intm.update...
matrix_mul.compute_at...
matrix_mul.update...
out.bound...
```
Could you give me some hints on the thinking process?

**Comment by abadams** [comment]
Yeah, this is a complicated schedule. I think the best way to read a schedule is this: First compile_to_conceptual_stmt in html format and load it in a browser. Next we divide and conquer - take each compute_root block of code one by one. For each compute_root Func, first note what the sequence of splits and reorders has done to its loop nest. You should be able to line up the stmt html with the scheduling directives. Then identify where all the child Funcs compute_at that Func are being produced and see how that corresponds to their compute_at calls. Continue recursively if things are compute_at the compute_at Func. Then look at the inner loops for each Func and note how the vectorize and unroll calls achieved that. 

It may help to first comment out the vectorize and unrolls in the schedule to see what the loop nest structure is without them, because those scheduling directives erase for loops structure in the final stmt, and that can obscure things. But if you do this, note that vectorize(x, 8) or unroll(x, 8) is actually a split of extent 8 followed by vectorizing or unrolling the inner loop that results, so if you want to comment it out you may need to replace it with a split.

The thing that's going to be hard to grok in this schedule is the rfactor call. That rewrites the Func DAG to make the summation a two stage hierarchical summation instead of a single big summation.
